app/models/agents/s3_agent.rb

```diff
@@ -0,0 +1,206 @@
+          errors.add(:base, "The 'watch' option is required and must be set to 'true' or 'false'")
+        end
+      when 'write'
+        if options['filename'].blank?
+          errors.add(:base, "filename must be specified in 'write' mode")
+        end
+        if options['data_path'].blank?
+          errors.add(:base, "data_path must be specified in 'write' mode")
+        end
+      end
+    end
+
+    def validate_access_key_id
+      !!buckets
+    end
+
+    def validate_access_key_secret
+      !!buckets
+    end
+
+    def complete_bucket
+      (buckets || []).collect { |room| {text: room.name, id: room.name} }
+    end
+
+    def working?
+      checked_without_error?
+    end
+
+    def check
+      return if interpolated['mode'] != 'read'
+      contents = safely do
+                   get_bucket_contents
+                 end
+      if boolify(interpolated['watch'])
+        watch(contents)
+      else
+        contents.each do |key, _|
+          create_event payload: get_file_pointer(key)
+        end
+      end
+    end
+
+    def get_io(file)
+      client.get_object(bucket: interpolated['bucket'], key: file).body
+    end
+
+    def receive(incoming_events)
+      return if interpolated['mode'] != 'write'
+      incoming_events.each do |event|
+        safely do
+          mo = interpolated(event)
+          client.put_object(bucket: mo['bucket'], key: mo['filename'], body: Utils.value_at(event.payload, mo['data_path']))
+        end
+      end
+    end
+
+    private
+
+    def safely
+      yield
+    rescue Aws::S3::Errors::AccessDenied => e
+      error("Could not access '#{interpolated['bucket']}' #{e.class} #{e.message}")
+    rescue Aws::S3::Errors::ServiceError =>e
+      error("#{e.class}: #{e.message}")
+    end
+
+    def watch(contents)
+      if last_check_at.nil?
+        self.memory = contents
+        return
+      end
+
+      new_memory = contents.dup
+
+      memory.each do |key, etag|
+        if contents[key].blank?
+          create_event payload: get_file_pointer(key).merge(event_type: :removed)
+        elsif contents[key] != etag
+          create_event payload: get_file_pointer(key).merge(event_type: :modified)
+        end
+        contents.delete(key)
+      end
+      contents.each do |key, etag|
+        create_event payload: get_file_pointer(key).merge(event_type: :added)
+      end
+
+      self.memory = new_memory
+    end
+
+    def get_bucket_contents
+      contents = {}
+      client.list_objects(bucket: interpolated['bucket']).each do |response|
+        response.contents.each do |file|
+          contents[file.key] = file.etag
+        end
+      end
+      contents
+    end
+
+    def client
+      @client ||= Aws::S3::Client.new(credentials: Aws::Credentials.new(interpolated['access_key_id'], interpolated['access_key_secret']),
+                                      region: interpolated['region'])
+    end
+
+    def buckets(log = false)
+      client.list_buckets.buckets
```

Comment:
Currently all validations and completions are done in separate requests, I cached it in case we optimize that later.